server/src/main/java/org/apache/druid/server/coordinator/duty/CloneHistoricals.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.druid.server.coordinator.stats.Dimension;
+import org.apache.druid.server.coordinator.stats.RowKey;
+import org.apache.druid.server.coordinator.stats.Stats;
+import org.apache.druid.timeline.DataSegment;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Handles cloning of historicals. Given the historical to historical clone mappings, based on
+ * {@link CoordinatorDynamicConfig#getCloneServers()}, copies any segments load or unload requests from the source
+ * historical to the target historical.
+ */
+public class CloneHistoricals implements CoordinatorDuty
+{
+  private static final Logger log = new Logger(CloneHistoricals.class);
+  private final SegmentLoadQueueManager loadQueueManager;
+
+  public CloneHistoricals(SegmentLoadQueueManager loadQueueManager)
+  {
+    this.loadQueueManager = loadQueueManager;
+  }
+
+  @Override
+  public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
+  {
+    final Map<String, String> cloneServers = params.getCoordinatorDynamicConfig().getCloneServers();
+    final CoordinatorRunStats stats = params.getCoordinatorStats();
+
+    if (cloneServers.isEmpty()) {
+      // No servers to be cloned.
+      return params;
+    }
+
+    // Create a map of host to historical.
+    final Map<String, ServerHolder> historicalMap = params.getDruidCluster()
+                                                          .getHistoricals()
+                                                          .values()
+                                                          .stream()
+                                                          .flatMap(Collection::stream)
+                                                          .collect(Collectors.toMap(
+                                                              serverHolder -> serverHolder.getServer().getHost(),
+                                                              serverHolder -> serverHolder
+                                                          ));
+
+    for (Map.Entry<String, String> entry : cloneServers.entrySet()) {
+      log.debug("Handling cloning for mapping: [%s]", entry);
+
+      final String sourceHistoricalName = entry.getKey();
+      final ServerHolder sourceServer = historicalMap.get(sourceHistoricalName);
+
+      if (sourceServer == null) {
+        log.error(
+            "Could not process clone mapping[%s] as source historical[%s] does not exist.",
+            sourceHistoricalName,
+            entry
+        );
+        continue;
+      }
+
+      final String targetHistoricalName = entry.getValue();
+      final ServerHolder targetServer = historicalMap.get(targetHistoricalName);
+
+      if (targetServer == null) {
+        log.error(
+            "Could not process clone mapping[%s] as target historical[%s] does not exist.",
+            targetHistoricalName,
+            entry
+        );
+        continue;
+      }
```

Comment:
We can probably merge these conditions into one.

```suggestion
      final String targetHistoricalName = entry.getValue();
      final ServerHolder targetServer = historicalMap.get(targetHistoricalName);

      if (sourceServer == null || targetServer == null) {
        log.error(
            "Could not process clone mapping[%s] as historical[%s] does not exist.",
            entry,
            (sourceServer == null ? sourceServerName : targetHistoricalName)
        );
        continue;
      }
```